pkg/api/handlers/memorysets.go

```diff
@@ -0,0 +1,168 @@
+package handlers
+
+import (
+	"time"
+
+	"github.com/obot-platform/obot/apiclient/types"
+	"github.com/obot-platform/obot/pkg/api"
+	"github.com/obot-platform/obot/pkg/hash"
+	v1 "github.com/obot-platform/obot/pkg/storage/apis/obot.obot.ai/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type MemorySetHandler struct {
+}
+
+func NewMemorySetHandler() *MemorySetHandler {
+	return &MemorySetHandler{}
+}
+
+func (*MemorySetHandler) AddMemories(req api.Context) error {
+	var (
+		memorySet v1.MemorySet
+		memories  []types.Memory
+	)
+
+	thread, err := getThreadForScope(req)
+	if err != nil {
+		return err
+	}
+
+	if err := req.Read(&memories); err != nil {
+		return err
+	}
+
+	// Current time for new memories
+	currentTime := types.NewTime(time.Now())
+
+	// Assign IDs based on content hash and deduplicate
+	newMemories := make(map[string]types.Memory)
+	for i := range memories {
+		memory := memories[i]
+		// Generate ID from content hash if not provided
+		if memory.ID == "" {
+			// Create a shorter URL-friendly hash (first 12 chars of SHA-256)
+			fullHash := hash.String(memory.Content)
+			memory.ID = fullHash[:12]
+		}
+		// Always set creation time to current time
+		memory.CreatedAt = *currentTime
+		newMemories[memory.ID] = memory
+	}
+
+	if err := req.Get(&memorySet, thread.Name); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+
+		// Create a new MemorySet
+		memorySet = v1.MemorySet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      thread.Name,
+				Namespace: req.Namespace(),
+			},
+			Spec: v1.MemorySetSpec{
+				ThreadName: thread.Name,
+				Manifest: types.MemorySetManifest{
+					Memories: []types.Memory{},
+				},
+			},
+		}
+
+		for _, memory := range newMemories {
+			memorySet.Spec.Manifest.Memories = append(memorySet.Spec.Manifest.Memories, memory)
+		}
+
+		if err := req.Create(&memorySet); err != nil {
+			return err
+		}
+	} else {
+		// Add memories to the existing MemorySet
+		existingMap := make(map[string]struct{})
+		for _, memory := range memorySet.Spec.Manifest.Memories {
+			existingMap[memory.ID] = struct{}{}
+		}
+
+		for id, memory := range newMemories {
+			if _, exists := existingMap[id]; exists {
+				continue
```

Comment:
Do we want to update existing memories? Like, make sure the content is updated?

Reply:
the id's are a hash of the content, so there shouldn't be a case where the content needs to be updated, but I can add that logic for robustness if you'd like

Reply:
I think the point is that this is an "add memory" endpoint and not an "update memories" endpoint, so I think this is good.